staging/vhost-device-gpu/src/virt_gpu.rs

```diff
@@ -330,16 +341,36 @@ impl VirtioGpu {
     }
 
     /// If the resource is the scanout resource, flush it to the display.
-    pub fn flush_resource(&mut self, resource_id: u32) -> VirtioGpuResult {
+    pub fn flush_resource(&mut self, resource_id: u32, gpu_backend: &mut GpuBackend, mem: &GuestMemoryMmap) -> VirtioGpuResult {
         if resource_id == 0 {
             return Ok(OkNoData);
         }
 
-        #[cfg(windows)]
-        match self.rutabaga.resource_flush(resource_id) {
-            Ok(_) => return Ok(OkNoData),
-            Err(RutabagaError::Unsupported) => {}
-            Err(e) => return Err(ErrRutabaga(e)),
+        let resource = self.resources.get(&resource_id).unwrap();
+        if resource.scanout_width != 0 && resource.scanout_height != 0 {
+            debug!("flush resource with associated scanout: {}x{}",resource.scanout_width, resource.scanout_height);
+            // lets assume hardcoded XRGB
+            const BYTES_PER_PIXEL: usize = 4;
+            let size = BYTES_PER_PIXEL * resource.scanout_width as usize * resource.scanout_height as usize;
```

Comment:
question: how did you arrive at this calculation?

Reply:
I currently don't check the format given in the `ResourceCreate2d` gpu command when creating the resource that will be used for the scanout. I assume it is `PIXMAN_x8r8g8b8`.  I don't think this has to be the case though, in which case the image would have to be converted here. The format of `VhostUserGpuUpdate` has to be `PIXMAN_x8r8g8b8`.

This could also maybe explain the crashes you have with out-of-bounds vector access.
Other thing that could be broken: did you by any chance resize the display window?
(maybe the widht/height is temporarily not correct in that case while resizing)

Not sure if we want to the bounds checking you are sugesting, if we have to do bounds checking like that, some of my assumptions is wrong, the problem is probably elsewhere (maybe the format, maybe something else I am unaware of).

Also note I tested this with a fixed resolution: `-device vhost-user-vga,chardev=char0,xres=640,yres=480`

Reply:
> I currently don't check the format given in the ResourceCreate2d gpu command when creating the > resource that will be used for the scanout. I assume it is PIXMAN_x8r8g8b8. I don't think this has to > be the case though, in which case the image would have to be converted here. The format of 
> VhostUserGpuUpdate has to be PIXMAN_x8r8g8b8.

hummm, instead of assuming the format maybe we should try to get the actual format. 

> This could also maybe explain the crashes you have with out-of-bounds vector access.
> Other thing that could be broken: did you by any chance resize the display window?
> (maybe the widht/height is temporarily not correct in that case while resizing)

> This could also maybe explain the crashes you have with out-of-bounds vector access.
> Other thing that could be broken: did you by any chance resize the display window?
> (maybe the widht/height is temporarily not correct in that case while resizing)

The crashes happen without me having to do anything (no resizing of window), I just start qemu and after few seconds it crashes. I didn't test this with a fixed resolution, the suggestion i made fixes the problem. also I don't think we should depend on fixed resolution.

Reply:
Yes totally we shouldn't depend on these things, but that's why this is a Draft - it is not really finished.

Even if we add the bounds checking, I think we should still investigate why the width and height dimensions are not correct. Maybe `scanout_width` and `scanout_height` need to be updated somewhere/under some condition we are not handling correctly.

Reply:
I don't think this has anything to do with the scanout width and height, the scanout width and height can vary depending on if you zoom in or out, does not mean that the dimensions are not correct. if there's a value that need to be re-evaluated i think it's this hardcoded constant `BYTES_PER_PIXEL`

Reply:
BYTES_PER_PIXEL, is definitely problematic. But I am not sure about `scanout_width` and `scanout_height` either.